test-app/tests/integration/components/headless-form-native-validation-test.gts

```diff
@@ -0,0 +1,754 @@
+            @name="firstName"
+            @validate={{fieldValidateCallback}}
+            as |field|
+          >
+            <field.label>First Name</field.label>
+            <field.input required pattern="^[A-Za-z]+$" data-test-first-name />
+            <field.errors data-test-first-name-errors as |errors|>
+              {{#each errors as |e index|}}
+                <div data-test-error={{index}}>
+                  <div data-test-error-type>
+                    {{e.type}}
+                  </div>
+                  <div data-test-error-value>
+                    {{e.value}}
+                  </div>
+                  <div data-test-error-message>
+                    {{e.message}}
+                  </div>
+                </div>
+              {{/each}}
+            </field.errors>
+          </form.field>
+          <form.field @name="lastName" as |field|>
+            <field.label>Last Name</field.label>
+            <field.input data-test-last-name />
+            <field.errors data-test-last-name-errors />
+          </form.field>
+          <button type="submit" data-test-submit>Submit</button>
+        </HeadlessForm>
+      </template>);
+
+      await click('[data-test-submit]');
+
+      assert
+        .dom('[data-test-first-name-errors] [data-test-error]')
+        .exists({ count: 3 });
+
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-type]'
+        )
+        .hasText('native');
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-value]'
+        )
+        .hasText('foo123');
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-message]'
+        )
+        .hasAnyText(); // validation error message is browser and locale dependant, so testing against actual message would be very brittle.
+
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="1"] [data-test-error-type]'
+        )
+        .hasText('uppercase');
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="1"] [data-test-error-value]'
+        )
+        .hasText('foo123');
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="1"] [data-test-error-message]'
+        )
+        .hasText('First name must be upper case!');
+
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="2"] [data-test-error-type]'
+        )
+        .hasText('notFoo');
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="2"] [data-test-error-value]'
+        )
+        .hasText('foo123');
+      assert
+        .dom(
+          '[data-test-first-name-errors] [data-test-error="2"] [data-test-error-message]'
+        )
+        .hasText('Foo is an invalid first name!');
+
+      assert.dom('[data-test-last-name-errors]').doesNotExist();
```

Comment:
Suggestion, nit: this test only seems to test for `errors` and not the success state?

Reply:
Good point, I'll add another one for this!